library/std/src/thread/mod.rs

```diff
@@ -717,6 +717,13 @@ where
 
 /// Gets a handle to the thread that invokes it.
 ///
+/// # Panics
+///
+/// Panics if called beyond the end of `main`, when the Rust standard library's
+/// thread state has been torn down. For example, use of `libc::atexit` on some
+/// platforms can hit this. In general much of the standard library is not okay
+/// to use before or after `main`.
```

Comment:
This isn't really the right place to mention this, as ChrisDenton said it would be good to have some module documentation (perhaps on the std top level module) explaining this and then linking to that section.

Reply:
I'll look for a spot. Is it fair to say 100% of libcore is fine to use before/after main, but the default assumption for libstd stuff like std::thread and std::io is it's UB to use before/after main, with the observable behavior being indistinguishable from a panic in the best case? How about liballoc?

Reply:
Some core and alloc functions could get poisoned through hooks like the global allocator or panic hooks.

> I'll look for a spot. 

Since this affects multiple modules maybe the top level?

> the default assumption for libstd stuff like std::thread and std::io is it's UB to use before/after main, with the observable behavior being indistinguishable from a panic in the best case

That's a tough one, I guess it's true that there might be UB lurking somewhere but feels a bit like a "known to the state of california to cause cancer" statement.

Reply:
> with the observable behavior being indistinguishable from a panic in the best case

The best case is that it just works. But that depends on the platform and operation.